### PR TITLE
ML: Adds a labeler script with various tools

### PR DESCRIPTION
**Pull request by mattlangford**
This is the start of the machine learning framework for use on the sub.
This tool is an easy to use tool for segmenting or labeling images. Right now there are three tools and two so-called image providers (more to come of both!)

**Tools:**
- Paint brush tool: paints the mask like a brush
- Polygon tool: Lets you draw polygons
- Move tool: Lets you move the mask around

**Image Providers:**
- Random color: mostly used for testing
- Webcam: gives you frame by frame access to your webcam

The goal with this is to be an easy to improve framework for segmenting images that will soon be used to train a learned segmenter. The requirements for adding new tools or image providers are fairly straight forward and just require simple adaptions to the current implementation.

P.S. if you find any little bugs when using it please fix them

### CODE REVIEW

File: machine_learning/CMakeLists.txt
Comment (kev-the-dev) [inline machine_learning/CMakeLists.txt]: Do you need to uncomment this for python to work?

File: machine_learning/labeling/image_providers.py
Comment (kev-the-dev) [inline machine_learning/labeling/image_providers.py]: might be nice to kwargs this too

**Review by kev-the-dev — CHANGES_REQUESTED** [review]
Really helpful package. Please add setup.py and the catkin python setup thing so it's rosified

File: machine_learning/CMakeLists.txt
Comment (mattlangford) [reply machine_learning/CMakeLists.txt]: I wasn't planning on exposing any of the inner workings to the system, therefore it wouldn't need to be there.
The image provider class could be helpful externally, but I don't think the machine learning folder is where it should live. Until there's a nice home for it set up (maybe in `mil_dev_tools`), it should stay as a private file only used by the labeler.

File: machine_learning/labeling/image_providers.py
Comment (mattlangford) [reply machine_learning/labeling/image_providers.py]: In reference to the above comment, only the labeler uses this function right now, so no need until it's exposed.

File: machine_learning/CMakeLists.txt
Comment (DSsoto) [inline machine_learning/CMakeLists.txt]: Can you remove commented out boilerplate?

File: machine_learning/labeling/image_providers.py
Comment (DSsoto) [inline machine_learning/labeling/image_providers.py]: I'm not seeing this being used anywhere, should it be a comment?

File: machine_learning/labeling/labeler.py
Comment (DSsoto) [inline machine_learning/labeling/labeler.py]: Is this a docstring for the whole labeler.py module?

File: machine_learning/labeling/labeler.py
Comment (DSsoto) [inline machine_learning/labeling/labeler.py]: Like what you did there.

File: machine_learning/package.xml
Comment (DSsoto) [inline machine_learning/package.xml]: I think that just like with sub and boat, we should establish a  mil namespace. I propose that all packages in this repo start with `mil_`. What say you?

**Review by DSsoto — CHANGES_REQUESTED** [review]
Requesting either package name change or good reasons not to, plus minor comments, questions.

Also have you considered making this a qt app?

File: machine_learning/package.xml
Comment (mattlangford) [reply machine_learning/package.xml]: Sure

File: machine_learning/CMakeLists.txt
Comment (mattlangford) [reply machine_learning/CMakeLists.txt]: I like to leave in the lines that could possibly be used in the future, that way if someone adds a c++ file or wants messages, they can just uncomment the lines rather then going to some other cmake file and copying what they need

File: machine_learning/labeling/image_providers.py
Comment (mattlangford) [reply machine_learning/labeling/image_providers.py]: Line 212 in the `labeler.py` file

File: machine_learning/labeling/labeler.py
Comment (mattlangford) [reply machine_learning/labeling/labeler.py]: I suppose it is

**Review by DSsoto — APPROVED** [review]

**Comment by DSsoto** [comment]
This is ready to merge but shouln't be until **__The Great Merge__** is complete and this is rebased on top of it.

**Comment by mattlangford** [comment]
Going to close based on some stuff. This'll still exist in my branch if we ever want it.

**Comment by DSsoto** [comment]
Wait how come? This seems like pretty useful stuff. I just wanted to wait until the great merge was done since this is lower priority.

**Comment by mattlangford** [comment]
@tessbianchi made some pretty compelling arguments to me about using an online labeling tool and adding infra to support that rather then rolling our own. The benefits of the online tool are too many to type so next time there is a call we can discuss.

**Comment by DSsoto** [comment]
Oh yea, she talked to me about that too. Fair enough.

**Pull request closed without merging**
